src/celeritas/grid/SplineXsCalculator.hh

```diff
@@ -0,0 +1,255 @@
+    }
+    else
+    {
+        true_low_idx = 0;
+    }
+    size_type true_high_idx
+        = min(lower_idx + order_steps + 1, loge_grid_.size());
+
+    // if the requested interpolation order is even, a direction must be
+    // selected to interpolate to
+    if (order_ % 2 == 0)
+    {
+        real_type low_dist = std::fabs(loge - loge_grid_[lower_idx]);
+        real_type high_dist = std::fabs(loge_grid_[lower_idx + 1] - loge);
+
+        // if we already clipped based on the bounding indices, don't clip
+        // again
+        if (true_high_idx - true_low_idx > order_ + 1)
+        {
+            if (low_dist > high_dist)
+            {
+                true_low_idx += 1;
+            }
+            else
+            {
+                true_high_idx -= 1;
+            }
+        }
+    }
+
+    return this->interpolate(energy.value(), true_low_idx, true_high_idx);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the cross section at the given index.
+ */
+CELER_FUNCTION real_type SplineXsCalculator::operator[](size_type index) const
+{
+    real_type energy = std::exp(loge_grid_[index]);
+    real_type result = this->get(index);
+
+    if (index >= data_.prime_index)
+    {
+        result /= energy;
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the raw cross section data at a particular index.
+ */
+CELER_FUNCTION real_type SplineXsCalculator::get(size_type index) const
+{
+    CELER_EXPECT(index < data_.value.size());
+    return reals_[data_.value[index]];
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the raw cross section data at a particular index.
+ */
+CELER_FUNCTION real_type SplineXsCalculator::interpolate(
+    real_type energy, size_type low_idx, size_type high_idx) const
+{
+    CELER_EXPECT(high_idx <= loge_grid_.size());
+    real_type result = 0.0;
+
+    // Outer loop over indices for contributing to the result
+    for (size_type outer_idx = low_idx; outer_idx < high_idx; ++outer_idx)
+    {
+        real_type outer_e = std::exp(loge_grid_[outer_idx]);
+        real_type num = 1.0;
+        real_type denom = 1.0;
+
+        // Inner loop over indices for determining the weight
+        for (size_type inner_idx = low_idx; inner_idx < high_idx; ++inner_idx)
+        {
+            // don't contribute for inner and outer index the same
+            if (inner_idx != outer_idx)
+            {
+                real_type inner_e = std::exp(loge_grid_[inner_idx]);
+                num *= (energy - inner_e);
+                denom *= (outer_e - inner_e);
+            }
+        }
+        real_type weight = num / denom;
+        real_type value = this->get(outer_idx);
+        if (outer_idx >= data_.prime_index)
+        {
+            value /= outer_e;
+        }
+
+        result += weight * value;
```

Comment:
Since registry pressure is so important for GPU, and I don't think the compiler can optimize this without `fast-math`, it might be better to combine weight and value into a single `weighted_value` scalar

Reply:
Makes sense, I have combined those values, let me know if you intended it to be done differently